File: ebpf/agent.c
// +build ignore

#include "vmlinux.h"
#include <bpf/bpf_helpers.h>
#include <bpf/bpf_tracing.h>
#include <bpf/bpf_core_read.h>

char __license[] SEC("license") = "Dual MIT/GPL";

#define EVENT_HTTP_500 1
#define EVENT_TCP_CONN 2

struct event_hdr {
    __u32 type;
};

struct http_event {
    struct event_hdr hdr;
    __u32 pid;
    __u32 tgid;
    __u8 snippet[32];
};

struct conn_event {
    struct event_hdr hdr;
    __u32 pid;
    __u32 tgid;
    __u32 daddr;
    __u16 dport;
    __u16 pad;
};

struct {
    __uint(type, BPF_MAP_TYPE_RINGBUF);
    __uint(max_entries, 1 << 20); // 1 MB
} events SEC(".maps");

// Force struct to be kept in the eBPF object
struct http_event *unused1 __attribute__((unused));
struct conn_event *unused2 __attribute__((unused));

SEC("tracepoint/syscalls/sys_enter_write")
int handle_sys_write(struct trace_event_raw_sys_enter *ctx) {
    __u64 id = bpf_get_current_pid_tgid();
    __u32 tgid = id >> 32;
    __u32 pid = id;

    const char *buf = (const char *)ctx->args[1];
    size_t count = (size_t)ctx->args[2];

    if (count < 12) {
        return 0;
    }

    char prefix[12];
    if (bpf_probe_read_user(&prefix, sizeof(prefix), buf) != 0) {
        return 0;
    }

    if (prefix[0] == 'H' && prefix[1] == 'T' && prefix[2] == 'T' && prefix[3] == 'P' &&
        prefix[4] == '/' && prefix[5] == '1' && prefix[6] == '.' &&
        prefix[8] == ' ' && prefix[9] == '5' && prefix[10] == '0' && prefix[11] == '0') {
        
        struct http_event *e;
        e = bpf_ringbuf_reserve(&events, sizeof(*e), 0);
        if (!e) {
            return 0;
        }

        e->hdr.type = EVENT_HTTP_500;
        e->pid = pid;
        e->tgid = tgid;
        
        bpf_probe_read_user(&e->snippet, sizeof(e->snippet), buf);
        bpf_ringbuf_submit(e, 0);
    }

    return 0;
}

SEC("tracepoint/syscalls/sys_enter_connect")
int handle_sys_connect(struct trace_event_raw_sys_enter *ctx) {
    __u64 id = bpf_get_current_pid_tgid();
    __u32 tgid = id >> 32;
    __u32 pid = id;

    struct sockaddr *addr = (struct sockaddr *)ctx->args[1];
    int addrlen = (int)ctx->args[2];

    if (addrlen != 16) { // sizeof(struct sockaddr_in)
        return 0;
    }

    short family = 0;
    if (bpf_probe_read_user(&family, sizeof(family), &addr->sa_family) != 0) {
        return 0;
    }

    if (family != 2) { // AF_INET
        return 0;
    }

    struct conn_event *e;
    e = bpf_ringbuf_reserve(&events, sizeof(*e), 0);
    if (!e) {
        return 0;
    }

    e->hdr.type = EVENT_TCP_CONN;
    e->pid = pid;
    e->tgid = tgid;
    
    // Read IP (offset 4) and Port (offset 2)
    bpf_probe_read_user(&e->dport, 2, ((char *)addr) + 2);
    bpf_probe_read_user(&e->daddr, 4, ((char *)addr) + 4);

    bpf_ringbuf_submit(e, 0);
    return 0;
}

File: ebpf/agent.go
//go:build linux
// +build linux

package ebpf

import (
	"bytes"
	"encoding/binary"
	"errors"
	"fmt"
	"log"
	"net"

	"github.com/cilium/ebpf/link"
	"github.com/cilium/ebpf/ringbuf"
	"github.com/cilium/ebpf/rlimit"
)

type EventType uint32

const (
	EventHTTP500 EventType = 1
	EventTCPConn EventType = 2
)

type EventHeader struct {
	Type EventType
}

type HTTPEvent struct {
	Header  EventHeader
	PID     uint32
	TGID    uint32
	Snippet [32]byte
}

type ConnEvent struct {
	Header EventHeader
	PID    uint32
	TGID   uint32
	DAddr  uint32
	DPort  uint16
	Pad    uint16
}

type Callbacks struct {
	OnHTTP500 func(pid uint32, snippet string)
	OnTCPConn func(pid uint32, destIP string, destPort uint16)
}

func StartAgent(cb Callbacks) error {
	if err := rlimit.RemoveMemlock(); err != nil {
		return err
	}

	objs := bpfObjects{}
	if err := loadBpfObjects(&objs, nil); err != nil {
		return err
	}

	lWrite, err := link.Tracepoint("syscalls", "sys_enter_write", objs.HandleSysWrite, nil)
	if err != nil {
		objs.Close()
		return fmt.Errorf("trace sys_enter_write: %v", err)
	}

	lConn, err := link.Tracepoint("syscalls", "sys_enter_connect", objs.HandleSysConnect, nil)
	if err != nil {
		lWrite.Close()
		objs.Close()
		return fmt.Errorf("trace sys_enter_connect: %v", err)
	}

	rd, err := ringbuf.NewReader(objs.Events)
	if err != nil {
		lConn.Close()
		lWrite.Close()
		objs.Close()
		return err
	}

	log.Println("[ebpf] Agent started, capturing HTTP 500s and Network Connections...")

	go func() {
		defer lWrite.Close()
		defer lConn.Close()
		defer objs.Close()
		defer rd.Close()

		for {
			record, err := rd.Read()
			if err != nil {
				if errors.Is(err, ringbuf.ErrClosed) {
					return
				}
				continue
			}

			var hdr EventHeader
			buf := bytes.NewBuffer(record.RawSample)
			if err := binary.Read(buf, binary.LittleEndian, &hdr); err != nil {
				continue
			}

			// Rewind buffer
			buf = bytes.NewBuffer(record.RawSample)

			if hdr.Type == EventHTTP500 {
				var ev HTTPEvent
				if err := binary.Read(buf, binary.LittleEndian, &ev); err != nil {
					continue
				}
				idx := bytes.IndexByte(ev.Snippet[:], 0)
				var snippetStr string
				if idx == -1 {
					snippetStr = string(ev.Snippet[:])
				} else {
					snippetStr = string(ev.Snippet[:idx])
				}
				if cb.OnHTTP500 != nil {
					cb.OnHTTP500(ev.PID, snippetStr)
				}
			} else if hdr.Type == EventTCPConn {
				var ev ConnEvent
				if err := binary.Read(buf, binary.LittleEndian, &ev); err != nil {
					continue
				}
				
				// Decode IP
				ipBytes := make([]byte, 4)
				binary.LittleEndian.PutUint32(ipBytes, ev.DAddr)
				destIP := net.IPv4(ipBytes[0], ipBytes[1], ipBytes[2], ipBytes[3]).String()
				
				// Decode Port (it's in network byte order big endian)
				portBytes := make([]byte, 2)
				binary.LittleEndian.PutUint16(portBytes, ev.DPort)
				destPort := binary.BigEndian.Uint16(portBytes)
				
				if cb.OnTCPConn != nil {
					cb.OnTCPConn(ev.PID, destIP, destPort)
				}
			}
		}
	}()

	return nil
}
